core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsAddressEndpointGroup.java

```diff
@@ -16,150 +16,183 @@
     }
 
-    /**
-     * Starts polling for service updates.
-     */
-    public void start() {
-        checkState(scheduledFuture == null, "already started");
-        scheduledFuture = eventLoop.scheduleAtFixedRate(this::query, 0, queryInterval.getSeconds(),
-                                                        TimeUnit.SECONDS);
+    private static List<DnsQuestion> newQuestions(String hostname, ResolvedAddressTypes resolvedAddressTypes) {
+        final ImmutableList.Builder<DnsQuestion> builder = ImmutableList.builder();
+        switch (resolvedAddressTypes) {
+            case IPV4_ONLY:
+            case IPV4_PREFERRED:
+            case IPV6_PREFERRED:
+                builder.add(new DefaultDnsQuestion(hostname, DnsRecordType.A));
+                break;
+        }
+        switch (resolvedAddressTypes) {
+            case IPV6_ONLY:
+            case IPV4_PREFERRED:
+            case IPV6_PREFERRED:
+            builder.add(new DefaultDnsQuestion(hostname, DnsRecordType.AAAA));
+            break;
+        }
+        return builder.build();
     }
 
-    /**
-     * Stops polling for service updates.
-     */
     @Override
-    public void close() {
-        if (scheduledFuture != null) {
-            scheduledFuture.cancel(true);
+    ImmutableSortedSet<Endpoint> onDnsRecords(List<DnsRecord> records, int ttl) throws Exception {
+        final ImmutableSortedSet.Builder<Endpoint> builder = ImmutableSortedSet.naturalOrder();
+        final boolean hasLoopbackARecords =
+                records.stream()
+                       .filter(r -> r instanceof DnsRawRecord)
+                       .map(r -> (DnsRawRecord) r)
+                       .anyMatch(r -> r.type() == DnsRecordType.A &&
+                                     r.content().getByte(r.content().readerIndex()) == 127);
+
+        for (DnsRecord r : records) {
+            if (!(r instanceof DnsRawRecord)) {
+                continue;
+            }
+
+            final DnsRecordType type = r.type();
+            final ByteBuf content = ((ByteBufHolder) r).content();
+            final int contentLen = content.readableBytes();
+
+            // Skip invalid records.
+            if (type == DnsRecordType.A) {
+                if (contentLen != 4) {
+                    warnInvalidRecord(DnsRecordType.A, content);
+                    continue;
+                }
+            } else if (type == DnsRecordType.AAAA) {
+                if (contentLen != 16) {
+                    warnInvalidRecord(DnsRecordType.AAAA, content);
+                    continue;
+                }
+            } else {
+                continue;
+            }
+
+            // Convert the content into an IP address into an endpoint.
+            final String ipAddr;
+            final byte[] addrBytes = new byte[contentLen];
+            content.getBytes(content.readerIndex(), addrBytes);
+
+            if (contentLen == 16) {
+                // 'result' contains both A and AAAA records.
```

Comment:
There isn't any variable named `result` here - do you mean `records may contain both A and AAAA records.`?

Reply:
Oops, just a cruft I forgot to remove during refactoring. :-)